### PR TITLE
Model Creation and Relationship 49970

### PR DESCRIPTION
**Pull request by abdullahallmehedi**

### CODE REVIEW

**Comment by Rownaknira** [comment]
ok

**Comment by manazirahsan** [comment]
okey

**Comment by noor-framgia** [comment]
ok

File: Gemfile [hunk 1]
```diff
@@ -1,47 +1,48 @@
...
 
 # Use Capistrano for deployment
-# gem 'capistrano-rails', group: :development
+# gem "capistrano-rails", group: :development
 
 group :development, :test do
-  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
-  gem 'byebug'
+  # Call "byebug" anywhere in the code to stop execution and get a debugger console
+  gem "byebug"
```
Comment (dieunb) [inline Gemfile]: use `pry` instead of it

File: app/models/book_image.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class BookImage < ActiveRecord::Base
+  belongs_to :book
+  enum type: [:cover, :thumbnail]
```
Comment (dieunb) [inline app/models/book_image.rb]: Should implement `carrierwave`

File: app/models/activity.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Activity < ActiveRecord::Base
+  belongs_to :user
+  has_many :likes
+  enum type: [:follow, :unfollow, :reading, :read, :favorite, :write_review,
```
Comment (dieunb) [inline app/models/activity.rb]: for `type` is same mysql keyword, do you think about `action`

File: app/models/book_image.rb (hunk 2 above)
Comment (dieunb) [inline app/models/book_image.rb]: Sam

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class Like < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :activity
+  enum type: [:like, :unlike]
```
Comment (dieunb) [inline app/models/like.rb]: do you think about `state`

File: app/models/relationship.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ActiveRecord::Base
+  belongs_to :follower, class_name: "User"
```
Comment (dieunb) [inline app/models/relationship.rb]: `"User" => User.name`

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+class User < ActiveRecord::Base
+  has_many :users_books, dependent: :destroy
+  has_many :reviews, dependent: :destroy
+  has_many :requests, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :activities, dependent: :destroy
+  has_many :likes, dependent: :destroy
+  has_many :books, through: :users_books
+  has_many :active_relationships, class_name:  "Relationship",
```
Comment (dieunb) [inline app/models/user.rb]: ` "Relationship" =>  Relationship.name`

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+                                  foreign_key: "follower_id",
+                                  dependent:   :destroy
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :passive_relationships, class_name:  "Relationship",
```
Comment (dieunb) [inline app/models/user.rb]: ` "Relationship" =>  Relationship.name`

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+class User < ActiveRecord::Base
+  has_many :users_books, dependent: :destroy
+  has_many :reviews, dependent: :destroy
+  has_many :requests, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :activities, dependent: :destroy
+  has_many :likes, dependent: :destroy
+  has_many :books, through: :users_books
+  has_many :active_relationships, class_name:  "Relationship",
+                                  foreign_key: "follower_id",
```
Comment (dieunb) [inline app/models/user.rb]:
```
"follower_id" => :follower_id
```

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+                                  dependent:   :destroy
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :passive_relationships, class_name:  "Relationship",
+                                   foreign_key: "followed_id",
```
Comment (dieunb) [inline app/models/user.rb]:
```
"followed_id" => :followed_id
```

File: app/models/review.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class Review < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :book
+  has_many :comments, dependent: :destroy
+  enum rating: [:worst, :bad, :satisfactory, :good, :best]
```
Comment (dieunb) [inline app/models/review.rb]: I think for rating, should use point number (1, 2,3,4,5)

File: .gitignore [hunk 11]
```diff
@@ -1,6 +1,8 @@
 *.rbc
 *.sassc
 .sass-cache
+/.gitignore
+/Gemfile.lock
```
Comment (dieunb) [inline .gitignore]: don't ignore Gemfile.lock

File: app/models/review.rb [hunk 12]
```diff
@@ -0,0 +1,6 @@
+class Review < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :book
+  has_many :comments, dependent: :destroy
+  enum rating: [1, 2, 3, 4, 5]
```
Comment (dieunb) [inline app/models/review.rb]: is necessary with enum rating?

File: app/models/relationship.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ActiveRecord::Base
+  belongs_to :follower, class_name: User.name
+  belongs_to :followed, class_name: User.name
+  validates :follower_id, presence: true
```
Comment (dieunb) [inline app/models/relationship.rb]: Do you think about `validates :follower`? And difference between them? :D

File: app/models/relationship.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
...
+  validates :followed_id, presence: true
```
Comment (dieunb) [inline app/models/relationship.rb]: Same

File: db/migrate/20151230075346_create_book_images.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+class CreateBookImages < ActiveRecord::Migration
+  def change
+    create_table :book_images do |t|
+      t.string :image
```
Comment (dieunb) [inline db/migrate/20151230075346_create_book_images.rb]: i think this field no need

File: db/migrate/20151230075346_create_book_images.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+      t.integer :type
```
Comment (dieunb) [inline db/migrate/20151230075346_create_book_images.rb]: i think this field no need, try implement carrierwave to know it

**Pull request merged**